pkg/filters/oidcadaptor/oidcadaptor.go

```diff
@@ -0,0 +1,428 @@
+	DeviceAuthorizationEndpoint       string   `json:"device_authorization_endpoint"`
+	TokenEndpoint                     string   `json:"token_endpoint"`
+	UserinfoEndpoint                  string   `json:"userinfo_endpoint"`
+	RevocationEndpoint                string   `json:"revocation_endpoint"`
+	JwksUri                           string   `json:"jwks_uri"`
+	ResponseTypesSupported            []string `json:"response_types_supported"`
+	SubjectTypesSupported             []string `json:"subject_types_supported"`
+	IdTokenSigningAlgValuesSupported  []string `json:"id_token_signing_alg_values_supported"`
+	ScopesSupported                   []string `json:"scopes_supported"`
+	TokenEndpointAuthMethodsSupported []string `json:"token_endpoint_auth_methods_supported"`
+	ClaimsSupported                   []string `json:"claims_supported"`
+	CodeChallengeMethodsSupported     []string `json:"code_challenge_methods_supported"`
+	GrantTypesSupported               []string `json:"grant_types_supported"`
+}
+
+type oidcIDToken struct {
+	AccessToken  string `json:"access_token"`
+	TokenType    string `json:"token_type"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	IdToken      string `json:"id_token"`
+}
+
+func init() {
+	filters.Register(kind)
+}
+
+func (o *Filter) Name() string {
+	return o.spec.Name()
+}
+
+func (o *Filter) Spec() filters.Spec {
+	return o.spec
+}
+
+func (o *Filter) Kind() *filters.Kind {
+	return kind
+}
+
+func (o *Filter) Init() {
+	//for testable delegate store interface operation to itself
+	o.store = o
+	if len(o.spec.Discovery) > 0 {
+		o.initDiscoveryOidcConf()
+	} else {
+		o.spec.oidcConfig = &oidcConfig{
+			AuthorizationEndpoint: o.spec.AuthorizationEndpoint,
+			TokenEndpoint:         o.spec.TokenEndpoint,
+			UserinfoEndpoint:      o.spec.UserinfoEndpoint,
+		}
+	}
+	parsed, err := url.Parse(o.spec.RedirectURI)
+	if err != nil {
+		logger.Errorf("parse redirectURI error: %s", err)
+	}
+	o.spec.redirectPath = parsed.Path
+}
+
+func (o *Filter) Inherit(previousGeneration filters.Filter) {
+	o.Init()
+	previousGeneration.Close()
+}
+
+func (o *Filter) Handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	var rw *httpprot.Response
+	if rw, _ = ctx.GetOutputResponse().(*httpprot.Response); rw == nil {
+		rw, _ = httpprot.NewResponse(nil)
+		ctx.SetOutputResponse(rw)
+	}
+	spec := o.spec
+	oidcConf := spec.oidcConfig
+	cookieExists := func() bool {
+		for _, c := range req.Cookies() {
+			if spec.CookieName == c.Name {
+				return true
+			}
+		}
+		return false
+	}
+	bearerTokenExists := func() bool {
+		var prefix = "Bearer"
+		authHeader := req.HTTPHeader().Get("Authorization")
+		if strings.HasPrefix(authHeader, prefix) && authHeader[len(prefix):] != "" {
+			return true
+		}
+		return false
+	}
+	if len(spec.CookieName) != 0 && cookieExists() {
+		return ""
+	}
+	if bearerTokenExists() {
+		return ""
+	}
```

Comment:
propose not using the closure functions, they make code hard to read in this case and can be avoid.